hax/hax/motr/planner.py

```diff
@@ -0,0 +1,297 @@
+    next_group_commands: Set[Type[BaseMessage]]
+    #
+    # If WorkPlanner is in shutting down mode.
+    # Shutting down mode is a special mode when WorkPlanner starts issuing
+    # poison pills ('Die' commands) making sure that all the threads get such
+    # commands ASAP.
+    #
+    # The flag is set to true by invoking WorkPlanner.shutdown() method.
+    is_shutdown: bool
+
+
+class WorkPlanner:
+    '''
+    Thread synchronizing block that is used as a work planner for Motr-aware
+    threads (see ConsumerThread). This synchronizing primitive guarantees that
+
+    1. The messages can be processed by an arbitrary number of ConsumerThread
+       threads.
+
+    2. The parallelism doesn't break semantics (some messages can be processed
+       in parallel to others, while some other not; the ConsumerThread's don't
+       need to worry about that).
+    '''
+    def __init__(self,
+                 init_state_factory: Optional[Callable[[], State]] = None):
+        fn = init_state_factory or self._create_initial_state
+
+        self.state = fn()
+        self.backlog: Deque[BaseMessage] = deque()
+        self.b_lock = Condition()
+
+    def is_empty(self) -> bool:
+        '''Checks whether the backlog is empty. Blocking call.'''
+        with self.b_lock:
+            return not self.backlog
+
+    def add_command(self, command: BaseMessage) -> None:
+        '''Adds the given command to the execution plan. Blocking call.'''
+        LOG.log(TRACE, '[WP]Before add_command: %s', command)
+        with self.b_lock:
+            cmd = self._assign_group(command)
+            LOG.log(TRACE, '[WP]Cmd %s is added. Current state: %s', cmd,
+                    self.state)
+            self.backlog.append(cmd)
+            # Some threads may be waiting because of an empty backlog - let's
+            # notify them
+            self.b_lock.notifyAll()
+
+    def _create_initial_state(self) -> State:
+        '''Default factory method that returns initial state.
+           Invoked from WorkPlanner's __init__ method.
+        '''
+        return State(next_group_id=0,
+                     active_commands=LinkedList(),
+                     taken_commands=LinkedList(),
+                     current_group_id=0,
+                     next_group_commands=set(),
+                     is_shutdown=False)
+
+    def _create_poison(self) -> BaseMessage:
+        '''Creates poison pill - Die command. Used in a special 'shutting down'
+           mode to stop worker threads gracefully.
+        '''
+
+        cmd = Die()
+        # Since it is a special case, no _assign_group() will be invoked to
+        # find the proper group. In fact, we want to stop the threads as soon
+        # as possible, that's why it doesn't make sense to postpone the
+        # command. In other words, it must belong to the group which is
+        # currently active.
+        cmd.group = self.state.current_group_id
```

Comment:
Can current_group_id change while the Die command is being assigned the group?
How about creating a group0 that can execute such high priority commands?

Reply:
Maintaining two different groups with different priorities looks like a complication to me and there is a risk to introduce a bug.

The case you're describing is done in a bit different way. 
1. We don't assume that `Die` commands will be put to WorkPlanner from outside (it is not hardly restricted though)
2. Instead, WorkPlanner will start generating `Die` commands automatically when it is  set to "shutting down" mode (see `shutdown()` method)
3. In "shutting down" mode WorkPlanner will always return `Die` command instance any time `get_next_command()` is invoked (although the backlog doesn't have to contain it). Returned `Die` command instance will always belong to the currently active group (reason: no need to wait, as the whole application is shutting down).
4. In "shutting down" mode WorkPlanner will always return `Die` command instance any time `ensure_allowed(cmd)` is invoked. In this case `ensure_allowed` works as if it "poisons" the given command turning it into `Die` (that is the reason why `ensure_allowed` has signature `BaseCommand -> BaseCommand` instead of `BaseCommand -> None`). Returned command also belongs to the current group.

Long story short, `Die` commands is a special case when groups stop making any sense. Effectively `Die` commands get executed as quick as possible ("on priority").

Reply:
Okay